docs/strategy-advanced.md

```diff
@@ -179,6 +179,45 @@ class AwesomeStrategy(IStrategy):
             return (-0.07 + current_profit)
         return 1
 ```
+#### Custom stoploss using an indicator from dataframe example
+
+Imagine you want to use `custom_stoploss()` to use a trailing indicator like e.g. "ATR"
+
+See: (Storing custom information using DatetimeIndex from `dataframe`
+)[WIP] on how to store the indicator into `custom_info`
+
+``` python
+from freqtrade.persistence import Trade
+from freqtrade.state import RunMode
+
+class AwesomeStrategy(IStrategy):
+
+    # ... populate_* methods
+
+    use_custom_stoploss = True
+
+    def custom_stoploss(self, pair: str, trade: 'Trade', current_time: datetime,
+                        current_rate: float, current_profit: float, **kwargs) -> float:
+
+        result = 1
+        if self.custom_info[pair] is not None and trade is not None:
+            # using current_time directly (like below) will only work in backtesting.
+            # so check "runmode" to make sure that it's only used in backtesting
+            if(self.dp.runmode == RunMode.BACKTEST):
+              relative_sl = self.custom_info[pair].loc[current_time]['atr]
+            # in live / dry-run, it'll be really the current time
+            else:
+              # but we can just use the last entry to get the current value
+              relative_sl =  self.custom_info[pair]['atr].iloc[ -1 ]
```

Comment:
for live, storing the dataframe is not really necessary, it's available from [`get_analyzed_dataframe()`](https://www.freqtrade.io/en/latest/strategy-customization/#get_analyzed_dataframepair-timeframe)

the problem is that this is not available during backtesting, which requires the "cust_info" workaround - which i'm ... hesitant to fully document fully - as it will raise memory requirements / problems - as the whole dataframe (or at least 2 columns of it) are stored and kept in memory.

Reply:
> as the whole dataframe (or at least 2 columns of it) are stored and kept in memory

for my usecase, spending a few megabytes or even gigabytes of system memory in backtesting, is a great tradeoff to designing a strategy which let's you ride a uptrend at zero risk (e.g. putting sl @ (atr) if profit+atr>1.5) or even having some kind of fixed risk/reward money management (e.g. getting intial_risk = open_rate+df[open_date]['atr']; take_profit = initial_risk*1.5+open_rate; if(current_rate>take_proift) #use trailng sl`) so I can compare W/L rate between strategies and still profit from trending.

I guess, what I want to say: there should be a note that it will increase memory consumption, for sure, yes, let's do that. make it even big and bold.

but the decision should still be made by the user. intentionally "hiding" it by not documenting it or giving a best practice example, only make people try to do it themself anyway. probably ignorante of best practices and the result will be something like my inital PR, that wouldn't even have worked in live.(which to me is a way worse result than using more memory)

backtesting on a device with limited memory like a raspi may be a bad idea anyways....

Reply:
> backtesting on a device with limited memory like a raspi may be a bad idea anyways....

agree on that - but not every user lives in a 1st world country - where buying a new pc every other year is possible - i've seen people use freqtrade on a 10yo pc without problem (ok, hyperopt may not run - but backtesting works just fine).

I think you misread my "hesitant" 
if i were against it, i'd not have commented on multiple lines but simply said "no" - otherwise i'd waste time for both of us.

The hesitant was more ... i knew about this possibility for a long time - but didn't document it intentionally ;) maybe should've used better wording...

Reply:
Ok, then let's do it!

Should I just add a comment warning with the same wording as the warning for not using too many indicators in `prepare_indicators()`?
Because it's basically like adding extra columns there, isn't it?

Or is it somehow more than that?

And regarding first world country or not, I think being hesitant adding more memory consumption is good, from a maintainer/Dev perspective, but like you said, most user are not really technical enough to appreciate the finer point of it. Post a warning, if they can't run it, they will have to triage their indicators/etc. 

Backtesting is a tool and if somebody wants to use it, because they think they can increase their alpha, the docs should at least point in the right direction, so they don't shoot themselves in the foot. 
If the tool then doesn't work with their hardware spec, they will either have to buy more, rethink their strategy or priorities what they want to do.

It's the same in all other areas. Nothing is stopping me to add thousands of indicators to use in `buy_trend()` except a note somewhere. But it's well documented how to do it if I want to. So I don't really get why `custom_stoploss` should be a different kind of animal (except it's newer).

But no need to decided this for the users beforehand.

 I'm not pounding on you with the same argument over and over, just trying to ease your mind about this. Because it seems a common issue in the Algo Trading space, that things that are bad from a technical view are still totally fine for the majority of user, which care most about getting their edge/alpha.

Reply:
> Or is it somehow more than that?

yes and no ... i'd have to look at it in detail - without `.copy()` it's for sure bearing the risk of causing a memory leak (keeping a reference around which shouldn't be).

Just put it as `!!! Warning` box ... it'll be fine i guess. I even think we had something similar for the cust_info before ... so 

> I'm not pounding on you with the same argument over and over

You're clear that we're now talking about details ... so the fact THAT we'll document this has been decided when you posted this PR ... 😆

Reply:
Yeah, still you are far more experienced with fretrades internal. Maybe you can spot a flaw in my logic if I lay it out.

At least it got me thinking about it in more detail. Also I made a note to look further into pythons memory management regarding references.

I don't think, with what I know from other languages, that copy a references will hurt it or make a memory leak. I was more afraid of side effects which get spaghetti very fast.

 There is probably a reference counter for each object and it will be deleted when not needed anymore. even if in this case RAII de-init is shortcuted. Also in which case would there be a IStrategy Model around still holding the df column in custom_info and no "global" dataframe?